### PR TITLE
Add lifecycle hooks to opensearch dashboard helm chart

### PR DESCRIPTION
**Pull request by bluehawk27**
### Description
[Describe what this change achieves]
Adds lifcycle hook that is currently supported in the helm template https://github.com/opensearch-project/helm-charts/blob/main/charts/opensearch-dashboards/templates/deployment.yaml#L140

### Issues Resolved
[List any issues this PR will resolve]
https://github.com/opensearch-project/helm-charts/issues/144
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: charts/opensearch-dashboards/values.yaml
Comment (TheAlgo) [inline charts/opensearch-dashboards/values.yaml:170]: Is this the best way to determine ?

File: charts/opensearch/CHANGELOG.md
Comment (TheAlgo) [inline charts/opensearch/CHANGELOG.md:22]: Looks like it got missed. We need to add the CHANGELOG

File: charts/opensearch-dashboards/values.yaml
Comment (bluehawk27) [reply charts/opensearch-dashboards/values.yaml:170]: I had just put it there as an example. If a working test is prefered. I can easily change it. @TheAlgo

Comment (TheAlgo) [reply charts/opensearch-dashboards/values.yaml:170]: Yes that will be great. Thanks @bluehawk27

File: charts/opensearch/CHANGELOG.md
Comment (bluehawk27) [reply charts/opensearch/CHANGELOG.md:22]: Im actually not 100% sure how these changes got pulled in. They appeared when I added my signature to the commit.

File: charts/opensearch-dashboards/values.yaml
Comment (peterzhuamazon) [reply charts/opensearch-dashboards/values.yaml:170]: Any progress @bluehawk27 ?

**Comment by peterzhuamazon** [comment]
Hi @bluehawk27 this is a change in dashboards but your version bump and change logs are modified in opensearch. Can you correct this in the corresponding files of opensearch-dashboards folder?

Thanks.

File: charts/opensearch-dashboards/values.yaml
Comment (DandyDeveloper) [inline charts/opensearch-dashboards/values.yaml:162]: Can we have something more formalized here? 

I'm spitballing but "Pod terminating. Beginning pod lifecycle commands.". Should we spit this out to `stderr`?

File: charts/opensearch-dashboards/values.yaml
Comment (DandyDeveloper) [inline charts/opensearch-dashboards/values.yaml:164]:
```suggestion
  postStart: {}
  #   exec:
```

Uncomment this and maybe add a comment above linking to the lifecycle page: https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks

File: charts/opensearch/CHANGELOG.md
Comment (DandyDeveloper) [reply charts/opensearch/CHANGELOG.md:22]: @bluehawk27 May be worthwhile bumping the chart again, as this is likely a merge in from main. 

But, we should also have a corresponding CHANGELOG section for your changes (per @TheAlgo statement).

**Review by DandyDeveloper — CHANGES_REQUESTED** [review]
Looks great, but I have some requests. 

Only other thing - We need to add the values for the lifecycle to the `README.md`

File: charts/opensearch-dashboards/values.yaml
Comment (bluehawk27) [reply charts/opensearch-dashboards/values.yaml:170]: @peterzhuamazon I'll knock this out tomorrow. I got swamped at work today.

File: charts/opensearch-dashboards/values.yaml
Comment (bluehawk27) [reply charts/opensearch-dashboards/values.yaml:164]: @DandyDeveloper I think it's good to document the usage as you have suggested with the link however I'd still leave it up to the user to make sure of the yaml object. The commented section is only there as an example

File: charts/opensearch-dashboards/values.yaml
Comment (bluehawk27) [reply charts/opensearch-dashboards/values.yaml:162]: @DandyDeveloper I just copy and pasted what is in the opensearch helm chart here any change should probably be done in both charts, and should be agreed upon.

Comment (bluehawk27) [reply charts/opensearch-dashboards/values.yaml:162]: @DandyDeveloper was the config section int the opensearch chart autogenerated? or hand rolled? https://github.com/opensearch-project/helm-charts/tree/main/charts/opensearch#configuration I notice that opensearch-dashboards is missing an equivalent.

Comment (TheAlgo) [reply charts/opensearch-dashboards/values.yaml:162]: @bluehawk27 This was hand rolled out. The config section in dashboards is pending. We have an issue open for this #129

**Comment by peterzhuamazon** [comment]
Seeing a lot of conflicts here.
Can you try to resolve them @bluehawk27 ?
Thanks.

**Comment by DandyDeveloper** [comment]
I neglected to follow up on this. If @bluehawk27 doesn't get a chance I'll fix the conflicts tomorrow.

**Comment by bluehawk27** [comment]
@DandyDeveloper @peterzhuamazon does every PR require a chart version bump? This PR doesn't materially change the chart behavior. It more or less just documents existing behavior.

**Comment by peterzhuamazon** [comment]
> @DandyDeveloper @peterzhuamazon does every PR require a chart version bump? This PR doesn't materially change the chart behavior. It more or less just documents existing behavior.

Yes @bluehawk27 
https://github.com/opensearch-project/helm-charts/pull/152#issuecomment-984559040

**Comment by bluehawk27** [comment]
ok thank you for answering @peterzhuamazon .  I can't keep resolving conflicts every time another PR gets merged in before this one so let me know when you plan on merging it in and I'll do a version bump.

**Comment by peterzhuamazon** [comment]
> ok thank you for answering @peterzhuamazon . I can't keep resolving conflicts every time another PR gets merged in before this one so let me know when you plan on merging it in and I'll do a version bump.

Reserve 1.5.8 for you @bluehawk27 we are also trying to improve the process, sorry about confusion.
We will merge your PR once #159 #84 #179 merged.

Thanks.

**Comment by peterzhuamazon** [comment]
@DandyDeveloper @TheAlgo please look at this PR and see if @bluehawk27 has resolved all issues.
Once you guys approve I will fix the versions and merge it.

Thanks.

**Comment by peterzhuamazon** [comment]
@DandyDeveloper @TheAlgo poke on this PR again as it sits for a while.
Could you guys check the code and see if we can proceed?
Ignore version conflicts.

Thanks.

**Comment by NybbleHub** [comment]
Hi guys,

Any news on this pull request ? Could be really good to have lifecycle block :)

Thanks !

**Comment by peterzhuamazon** [comment]
Adding @prudhvigodithi to take a look on this.

**Comment by TheAlgo** [comment]
Apologies @bluehawk27 , this got completely missed. Could you resolve the conversations for the comments you have addressed? We can take care of the conflicts and chart bumps

**Comment by bluehawk27** [comment]
@TheAlgo I resolved the convo's :) hope that helps move this forward. Cheers

**Comment by peterzhuamazon** [comment]
Hi @DandyDeveloper @TheAlgo @prudhvigodithi please review again, I will resolve the version conflicts once you guys are good.
Thanks.

**Comment by prudhvigodithi** [comment]
LGTM!

**Comment by peterzhuamazon** [comment]
Hi @bluehawk27 I have rebased your branch on main and remove the opensearch changelogs as only dashboards get changed with lifecycle hooks.

@TheAlgo @DandyDeveloper @prudhvigodithi please review.

Thanks.

**Comment by prudhvigodithi** [comment]
Hey @bluehawk27 I have done an upstream fetch and merge, to keep the versions in consistent with yesterday's release PR
https://github.com/opensearch-project/helm-charts/pull/283, additionally I have fixed some lint error.

**Review by prudhvigodithi — APPROVED** [review]

**Comment by prudhvigodithi** [comment]
LGTM @DandyDeveloper and @TheAlgo please review.
Thank you

**Review by TheAlgo — APPROVED** [review]
LGTM!

**Review by TheAlgo — COMMENTED** [review]
Can we add this option in the [README](https://github.com/opensearch-project/helm-charts/blob/main/charts/opensearch-dashboards/README.md) as well?

**Comment by TheAlgo** [comment]
Any update on this @bluehawk27 ?

**Review by DandyDeveloper — APPROVED** [review]

**Comment by peterzhuamazon** [comment]
I am going to add to readme and resolve conflict/backport @TheAlgo @prudhvigodithi

File: charts/opensearch-dashboards/README.md
Comment (prudhvigodithi) [inline charts/opensearch-dashboards/README.md:54]: @peterzhuamazon Thanks for adding this, can you please change this to See `values.yaml` for an example`

**Comment by peterzhuamazon** [comment]
Backport 1.x: https://github.com/opensearch-project/helm-charts/pull/305

Thanks @bluehawk27 for the contribution and sorry for the delay.

Thanks.

**Pull request merged**
